monolith/src/monolith.rs

```diff
@@ -0,0 +1,226 @@
+    pub lookup1: Vec<u16>,
+    pub lookup2: Vec<u16>,
+    pub mds: Mds,
+
+    _phantom: PhantomData<Mds>,
+}
+
+impl<Mds, const WIDTH: usize, const NUM_ROUNDS: usize> MonolithMersenne31<Mds, WIDTH, NUM_ROUNDS>
+where
+    Mds: MdsPermutation<Mersenne31, WIDTH>,
+{
+    pub const NUM_BARS: usize = 8;
+
+    pub fn new(mds: Mds) -> Self {
+        assert!(WIDTH >= 8);
+        assert!(WIDTH <= 24);
+        assert_eq!(WIDTH % 4, 0);
+
+        let round_constants = Self::instantiate_round_constants();
+        let lookup1 = Self::instantiate_lookup1();
+        let lookup2 = Self::instantiate_lookup2();
+
+        Self {
+            round_constants,
+            lookup1,
+            lookup2,
+            mds,
+            _phantom: PhantomData,
+        }
+    }
+
+    fn s_box(y: u8) -> u8 {
+        let y_rot_1 = (y >> 7) | (y << 1);
+        let y_rot_2 = (y >> 6) | (y << 2);
+        let y_rot_3 = (y >> 5) | (y << 3);
+
+        let tmp = y ^ !y_rot_1 & y_rot_2 & y_rot_3;
+        (tmp >> 7) | (tmp << 1)
+    }
+
+    pub fn final_s_box(y: u8) -> u8 {
+        debug_assert_eq!(y >> 7, 0); // must be a 7-bit value
+
+        let y_rot_1 = (y >> 6) | (y << 1);
+        let y_rot_2 = (y >> 5) | (y << 2);
+
+        let tmp = (y ^ !y_rot_1 & y_rot_2) & 0x7F;
+        ((tmp >> 6) | (tmp << 1)) & 0x7F
+    }
+
+    fn instantiate_lookup1() -> Vec<u16> {
+        (0..=u16::MAX)
+            .map(|i| {
+                let lo = (i >> 8) as u8;
+                let hi = i as u8;
+                ((Self::s_box(lo) as u16) << 8) | Self::s_box(hi) as u16
+            })
+            .collect()
+    }
+
+    fn instantiate_lookup2() -> Vec<u16> {
+        (0..(1 << 15))
+            .map(|i| {
+                let lo = (i >> 8) as u8;
+                let hi: u8 = i as u8;
+                ((Self::final_s_box(lo) as u16) << 8) | Self::s_box(hi) as u16
+            })
+            .collect()
+    }
+
+    fn random_field_element(shake: &mut Shake128Reader) -> Mersenne31 {
+        let mut val = get_random_u32(shake);
+        while val >= Mersenne31::ORDER_U32 {
+            val = get_random_u32(shake);
+        }
+
+        Mersenne31::from_canonical_u32(val)
+    }
+
+    fn init_shake() -> Shake128Reader {
+        let mut shake = Shake128::default();
+        shake.update("Monolith".as_bytes());
+        shake.update(&[WIDTH as u8, NUM_ROUNDS as u8]);
+        shake.update(&Mersenne31::ORDER_U32.to_le_bytes());
+        shake.update(&[8, 8, 8, 7]);
+        shake.finalize_xof()
+    }
+
+    fn instantiate_round_constants() -> Vec<[Mersenne31; WIDTH]> {
+        let mut shake = Self::init_shake();
+
+        vec![[Mersenne31::ZERO; WIDTH]; NUM_ROUNDS - 1]
+            .iter()
+            .map(|arr| arr.map(|_| Self::random_field_element(&mut shake)))
+            .collect()
+    }
+
+    pub fn concrete(
+        &self,
+        state: &mut [Mersenne31; WIDTH],
+        round_constants: Option<&[Mersenne31; WIDTH]>,
```

Comment:
It seems like the spec describes concrete as MDS only, with constant addition as its own step

> R_i(·) = c^(i) + Concrete ◦ Bricks ◦ Bars(·)

I think it would also be best to avoid the branch for performance reasons, and instead have a separate function that unconditionally adds constants (not called in the first round).